Enonic UI: Create a WorkflowContentIcon #9040

skjulestad opened the pull request:
Added WorkflowBadge.tsx
Slimmed down list-style
Wrapped ContentIcon.tsx with WorkflowBadge.tsx in:
ContentItem.tsx
ContentItemCheckable.tsx

skjulestad committed:
Enonic UI: Create a WorkflowContentIcon #9040

Added WorkflowBadge.tsx
Slimmed down list-style
Wrapped ContentIcon.tsx with WorkflowBadge.tsx in:
ContentItem.tsx
ContentItemCheckable.tsx

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentIcon.tsx:
```diff
@@ -78,7 +78,7 @@ export const ContentIcon = ({
         // 2x size for better quality
         return <Image
-            className={isImageType ? 'object-cover' : 'object-contain'}
+            className={`w-[24px] h-[24px] ${isImageType ? 'object-cover' : 'object-contain'}`}
```
No custom pixel classes. It's `w-6`. I doubt that  `w-[24px]` will be used elsewhere, but it'll create an additional class in resulted CSS file.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/WorkflowBadge.tsx:
```diff
@@ -0,0 +1,68 @@
+
+        status === WorkflowStateStatus.INVALID &&
+        'text-[var(--color-danger)] [&>circle]:fill-current [&>path]:stroke-white [&>path]:scale-[1.25]',
+    );
```
Ideal place to use `cva` here. Might need to install this dependency.

edloidas replied:
Since the icon can be moved to a separate component, `cva` can be applied there, not in this component.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/WorkflowBadge.tsx:
```diff
@@ -0,0 +1,68 @@
+        ? CircleCheck
+        : status === WorkflowStateStatus.IN_PROGRESS
+          ? CircleAlert
+          : CircleX;
```
Move to a separate private component. Could name `<StatusIcon>` or something like that. Please, don't use ternaries here, use switch.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/WorkflowBadge.tsx:
```diff
@@ -0,0 +1,68 @@
+          ? WorkflowStateStatus.READY
+          : summary.isInProgress()
+            ? WorkflowStateStatus.IN_PROGRESS
+            : null;
```
Move outside. Create a file `src/assets/js/app/util/content.ts` and name it `calcWorkflowStateStatus(summary: ContentSummary): WorkflowStateStatus`. Use if-else inside.
> What an awful name `WorkflowStateStatus`...

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/WorkflowBadge.tsx:
```diff
@@ -0,0 +1,68 @@
+                                  summary,
+                                  children,
+                                  overlaySize = 12,
+                              }: Props): JSX.Element | null {
```
You always return `JSX.Element`, not `null`.

alansemenov reviewed (commented):
As far as I can see, `overlaySize` already has a default value of 12, so I don't see why we need to pass it every time. Same with `size` for ContentIcon - we are showing it in hundreds of places in UI, so we should stick to default values without having to pass them every time.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/WorkflowBadge.tsx:
```diff
@@ -0,0 +1,68 @@
+
+type Props = {
+    summary?: ContentSummary | null;
+    children?: ReactNode;
```
- Accept `status` instead of `summary`. Don't make it optional, but allow `null`. 
- `children` must not be optional here, otherwise this component makes no sense.

We should have our props as accurate and minimal as possible. We also should avoid passing objects, especially created with classes, as they tend to change (the reference), while keeping the primitive inside intact. This will lead to unnecessary recalculations and redraws. Also hurt your memo from previous files.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentItemCheckable.tsx:
```diff
@@ -22,10 +23,15 @@ const ContentItemCheckableComponent = ({
     const label = String(content.getPath());
     const contentType = String(content.getType());
     const url = content.getContentSummary().getIconUrl();
+    const summary = content.getContentSummary();
```
Calculate status here and pass it instead:
```
const status = calcWorkflowStateStatus(content.getContentSummary());
```
> The details described in comments to other files.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/WorkflowBadge.tsx:
```diff
@@ -0,0 +1,68 @@
+      {children}
+            <span className="absolute -top-0.5 -right-0.5">
+        <Icon className={iconClass} strokeWidth={2} size={overlaySize} />
+      </span>
```
Should we really wrap it in `<span>`? I think it's okay to apply classes to the icon directly.

skjulestad replied:
Unavoidable.
I need root element for ContentIcon and statusIcon for classnames.

edloidas commented on modules/lib/src/main/resources/assets/js/app/ui2/list/ContentItemCheckable.tsx:
```diff
@@ -22,10 +23,15 @@ const ContentItemCheckableComponent = ({
+            <WorkflowBadge summary={summary} overlaySize={12}>
+                <ContentIcon contentType={contentType} url={url} size={24} />
+            </WorkflowBadge>
+        ),
```
I think that this one block will be reused more than once. Then there is a "rule": If the component used 3+ times, and they are coupled together (they both operate with content) and not just stateless UI blocks, then it OK to compose them in one component (usually keeping component separate and combine them in parent better).

- Let's move this logic to `<WorkflowContentIcon>`.
- Move `<WorkflowBadge>` and all other code there too. We don't need it as a separate component, just some inner code. I think you don't even need badge itself, you can use `<ContentIcon>` instead of abstract children directly after all those changes described bellow in other comments.

skjulestad committed:
Enonic UI: Create a WorkflowContentIcon #9040

Removed WorkflowBadge.tsx
Replaced with WorkflowContentIcon.tsx
Use ContentIcon.tsx directly within WorkflowContentIcon.tsx
Added util function content.ts

edloidas committed:
Enonic UI: Create a WorkflowContentIcon #9040
